pkg/commands/iamcleanup/iamcleanup.go

```diff
@@ -0,0 +1,187 @@
+func (c *IAMCleaner) Do(
+	ctx context.Context,
+	scope string,
+	iamQuery string,
+	evaluateCondition bool,
+) error {
+	logger := logging.FromContext(ctx)
+	iams, err := c.assetInventoryClient.IAM(ctx, scope, iamQuery)
+	if err != nil {
+		return fmt.Errorf("failed to get iam: %w", err)
+	}
+
+	logger.Debugw("got IAM", "number_of_iam_memberships", len(iams), "scope", scope, "query", iamQuery)
+
+	// We group by URI to confirm we only delete 1 ResourceID/Member combination at a single time.
+	// Multiple concurrent requests to delete IAM for a particular ResourceID/Member result in a conflict.
+	var iamsToDelete map[string][]*assetinventory.AssetIAM
+	if evaluateCondition {
+		iamsToDelete = groupByURI(filterByEvaluation(ctx, iams))
+	} else {
+		iamsToDelete = groupByURI(iams)
+	}
+
+	logger.Debugw("Cleaning up IAM", "number_of_iam_memberships_to_remove", countAll(iamsToDelete))
+
+	w := workerpool.New[*workerpool.Void](&workerpool.Config{
+		Concurrency: c.maxConcurrentRequests,
+		StopOnError: true,
+	})
+	for _, is := range iamsToDelete {
+		iamMemberships := is
+		if err := w.Do(ctx, func() (*workerpool.Void, error) {
+			for _, iamMembership := range iamMemberships {
+				switch iamMembership.ResourceType {
+				case assetinventory.Organization:
+					if err := c.iamClient.RemoveOrganizationIAM(ctx, iamMembership); err != nil {
+						return nil, fmt.Errorf("failed to remove org IAM: %w", err)
+					}
+				case assetinventory.Folder:
+					if err := c.iamClient.RemoveFolderIAM(ctx, iamMembership); err != nil {
+						return nil, fmt.Errorf("failed to remove folder IAM: %w", err)
+					}
+				case assetinventory.Project:
+					if err := c.iamClient.RemoveProjectIAM(ctx, iamMembership); err != nil {
+						return nil, fmt.Errorf("failed to remove project IAM: %w", err)
+					}
+				default:
+					return nil, fmt.Errorf("unable to remove membership for unsupported resource type %s", iamMembership.ResourceType)
+				}
+			}
+			return nil, nil
+		}); err != nil {
+			return fmt.Errorf("failed to execute remove IAM task: %w", err)
+		}
+	}
+	if _, err := w.Done(ctx); err != nil {
+		return fmt.Errorf("failed to execute IAM Removal tasks in parallel: %w", err)
+	}
+
+	logger.Debugw("Successfully deleted IAM", "number_of_iam_memberships_removed", len(iamsToDelete))
+
+	return nil
+}
+
+func filterByEvaluation(ctx context.Context, iams []*assetinventory.AssetIAM) []*assetinventory.AssetIAM {
+	logger := logging.FromContext(ctx)
+	var filteredResults []*assetinventory.AssetIAM
+	for _, i := range iams {
+		passed, err := evaluateIAMConditionExpression(ctx, i.Condition.Expression)
+		if err != nil {
+			logger.Warnw("Failed to parse expression (CEL) for IAM membership", "membership", i, "error", err)
+		} else if !*passed {
+			filteredResults = append(filteredResults, i)
+		}
+	}
+	return filteredResults
+}
+
+func evaluateIAMConditionExpression(ctx context.Context, expression string) (*bool, error) {
```

Comment:
Is this even possible? Some condition relies on request attributes that only exist in a real request, e.g. https://cloud.google.com/iam/docs/conditions-attribute-reference#ip-port

Reply:
I did think about this but decided to defer until we had more use cases. I could confirm that the condition ONLY uses `request.time` for the time being. wdyt?

Reply:
sgtm

Reply:
Done